tests/e2e-driver/internal/runner/runner.go

```diff
@@ -0,0 +1,180 @@
+	if err != nil {
+		return fmt.Errorf("failed to create temporary dir: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	tmpPkgPath := filepath.Join(tmpDir, r.pkgName)
+	// create result dir
+	resultsPath := filepath.Join(tmpDir, "results")
+	err = os.Mkdir(resultsPath, 0755)
+	if err != nil {
+		return fmt.Errorf("failed to create results dir %s: %w", resultsPath, err)
+	}
+
+	// copy package to temp directory
+	err = copyDir(r.pkgPath, tmpPkgPath)
+	if err != nil {
+		return fmt.Errorf("failed to copy package: %w", err)
+	}
+
+	// init and commit package files
+	err = r.preparePackage(tmpPkgPath)
+	if err != nil {
+		return fmt.Errorf("failed to prepare package: %w", err)
+	}
+
+	// run function
+	// TODO: change to pipeline when it's ready
+	kptArgs := []string{"fn", "run", tmpPkgPath, "--results-dir", resultsPath}
+	if r.network {
+		kptArgs = append(kptArgs, "--network")
+	}
+	o, err := runCommand("", "kpt", kptArgs)
+
+	// compare results
+	err = r.compareResult(err, tmpPkgPath, resultsPath)
+	if err != nil {
+		return fmt.Errorf("%w\nkpt output:\n%s", err, o)
+	}
+	return nil
+}
+
+func (r *Runner) preparePackage(pkgPath string) error {
+	err := gitInit(pkgPath)
+	if err != nil {
+		return err
+	}
+
+	err = gitAddAll(pkgPath)
+	if err != nil {
+		return err
+	}
+
+	return gitCommit(pkgPath, "first")
+}
+
+func (r *Runner) compareResult(exitErr error, tmpPkgPath, resultsPath string) error {
+	// get exit code
+	exitCode := 0
+	if e, ok := exitErr.(*exec.ExitError); ok {
+		exitCode = e.ExitCode()
+	} else if exitErr != nil {
+		return fmt.Errorf("cannot get exit code from %w", exitErr)
+	}
+	// compare exit code
+	b, err := ioutil.ReadFile(filepath.Join(tmpPkgPath, expectedDir, expectedExitCodeFile))
+	var expectedExitCode int
+	if os.IsNotExist(err) {
+		expectedExitCode = 0
+	} else if err != nil {
+		return fmt.Errorf("failed to read expected exit code: %w", err)
+	} else {
+		expectedExitCode, err = strconv.Atoi(strings.TrimSpace(string(b)))
+		if err != nil {
+			return fmt.Errorf("cannot convert exit code %s to int: %w", b, err)
+		}
+	}
+
+	if exitCode != expectedExitCode {
+		return fmt.Errorf("actual exit code %d doesn't match expected %d", exitCode, expectedExitCode)
+	}
+
+	if exitCode != 0 {
+		// compare results
+		l, err := ioutil.ReadDir(resultsPath)
+		if err != nil {
+			return fmt.Errorf("failed to get files in results dir: %w", err)
+		}
+		if len(l) != 1 {
+			return fmt.Errorf("unexpected results files number %d, should be 1", len(l))
+		}
+		resultsFile := l[0].Name()
+		actualResults, err := ioutil.ReadFile(filepath.Join(resultsPath, resultsFile))
+		if err != nil {
+			return fmt.Errorf("failed to read actual results: %w", err)
+		}
+		expectedResults, err := ioutil.ReadFile(filepath.Join(tmpPkgPath, expectedDir, expectedResultsFile))
+		if err != nil {
+			return fmt.Errorf("failed to read expected results: %w", err)
+		}
+		actual := strings.TrimSpace(string(actualResults))
```

Comment:
Why is this necessary?

Either:
1- we care to test the exact raw text of results YAML -> Compare files as is
2- we just want to compare the YAML content -> test the parsed content of the yaml instead of raw text

Given we already assert exact git diff for resources, then I think you want #1?

Reply:
Yes. The reason that I add `TrimSpace` here is to avoid false positive test failure caused by newline char in the end of expected files. Some users will add \n in the end of file while some will not.

Reply:
but why is the user editing the results file directly. Shouldn't be auto-generated by kpt?

Reply:
If we can guarantee that the results file will be copied (not the content) then we can avoid trimming spaces. And if someone open the expected results file with editor that will modify the `\n` in the file end (at least VS Code has such option, some git tool may do it as well) the difference between the expected and actual results is very hard to identify.